### ISSUE TITLE
useAsyncFn returned asyncFn should throw an error

### ISSUE THREAD

**Issue by gavinxgu** [streamich/react-use#981]
**What is the current behavior?**

When I call the wrapped asyncFn, the promise returned will not throw the error. However, sometimes it is common to do some sync jobs after an async job. What is the best practice for this situation?

[Reproduce Link](https://codesandbox.io/embed/festive-water-ocv1y?fontsize=14&hidenavigation=1&theme=dark)

**What is the expected behavior?**

Should throw an error in useAsyncFn

**A little about versions:**  
- _OS_:  MacOS 10.15.3
- _Browser (vendor and version)_: Chrome 79.0.3945.130
- _React_:  16.12.0
- _`react-use`_: v14.0
- _Did this worked in the previous package version?_ No

**Comment by whinc** [comment]
I don't think asyncFn should throw an error. The purpose of hooks is to update the UI status, and the success or failure of asynchronous function calls is reflected by the status. Throwing errors is imperative and may be you should use native JS function calls.

**Comment by chardydev** [comment]
I also agree that the promise should resolve with the error instead of returning it, for a couple of reasons:

1. Returning the error can make it difficult to disambiguate between the actual resolved object and a possibly unknown error type.
2. A callsite that invokes the the wrapped asyncFn absolutely may care if the promise rejected and may need to handle that appropriately.

Will you consider a PR for changing this behavior?

**Comment by pepsighan** [comment]
While I do believe the default behavior of useAsyncFn should be to not throw errors as @whinc said. But there are instances that I have faced where as @chardydev says in point 2, that the caller may want to distinguish whether an error was thrown to decide what actions to take after that.

I would propose modification to the options of the `useAsyncFn` or introduce a new hook altogether. The alternate hook/option would behave such that:
1. Errors and loading states work as in the current `useAsyncFn` (for the UI).
2. Errors should also be thrown so that the callers may catch it themselves (for the caller).

If, presuming that this is what the lib wants to pursue, we go via the options path, then this is what I was thinking it could look like:
```typescript
type UseAsyncFnOptions<T> = {
  // The default value to use before a value is resolved from the promise.
  defaultValue?: T;

  // If true, the call function will return the value after executing [fn] and rethrow
  // error if caught.
  doNotConsume?: boolean;
};
```

**Comment by s97712** [comment]
I think it shoud return `{ value?: T, error?: any}`  for the caller to handle the error correctly

**Comment by s97712** [comment]
> I think it shoud return `{ value?: T, error?: any}` for the caller to handle the error correctly

Typescript cannot infer that an error may be returned, which may lead to incorrect use of the return value, which did happen to me

**Comment by noinskit** [comment]
I just hit this as well! Returning an error object is breaking the contract in Typescript. As such, Typescript is actually preventing me from checking if the function returned an error, because it's not allowed to do that according to its signature.
